src/components/course/EnrollButton.jsx

```diff
@@ -1,210 +1,49 @@
-import { Button } from '@edx/paragon';
+import React, { useContext } from 'react';
+import { useLocation } from 'react-router-dom';
 
-import { UserSubsidyContext } from '../enterprise-user-subsidy';
+import EnrollButtonLabel from './enrollment/EnrollButtonLabel';
 import { CourseContext } from './CourseContextProvider';
-import EnrollButtonLabel from './EnrollButtonLabel';
-import EnrollModal from './EnrollModal';
-
-import { features } from '../../config';
-
-import { useCourseEnrollmentUrl } from './data/hooks';
-
-import {
-  hasCourseStarted,
-  findUserEnrollmentForCourse,
-  findHighestLevelSeatSku,
-  findOfferForCourse,
-  shouldUpgradeUserEnrollment,
-} from './data/utils';
-
-const EnrollButtonWrapper = ({
-  as: Component,
-  children,
-  ...props
-}) => (
-  <div className="enroll-wrapper" style={{ width: 270 }}>
-    <Component {...props}>
-      {children}
-    </Component>
-  </div>
-);
-
-EnrollButtonWrapper.propTypes = {
-  children: PropTypes.node.isRequired,
-  as: PropTypes.elementType,
-};
-
-EnrollButtonWrapper.defaultProps = {
-  as: Button,
-};
+import { useEnrollData, useSubsidyData } from './data/hooks';
+import EnrollAction from './enrollment/EnrollAction';
+import { determineEnrollmentType } from './enrollment/utils';
 
 export default function EnrollButton() {
-  const { enterpriseConfig, config } = useContext(AppContext);
   const { state: courseData } = useContext(CourseContext);
-  const { subscriptionLicense, offers: { offers, offersCount } } = useContext(UserSubsidyContext);
-  const location = useLocation();
-  const [isModalOpen, setIsModalOpen] = useState(false);
-
+  const { activeCourseRun, userEntitlements } = courseData;
   const {
+    isUserEnrolled, isEnrollable, isCourseStarted, userEnrollment,
+  } = useEnrollData();
+  const subsidyData = useSubsidyData({ location: useLocation() });
+  const { enrollmentUrl, subscriptionLicense } = subsidyData;
+
+  // decide which enrollment type applies (one of the various scenarios supported)
+  const enrollmentType = determineEnrollmentType({
     activeCourseRun,
-    userEnrollments,
     userEntitlements,
-    userSubsidy,
-    catalog: { catalogList },
-  } = courseData;
-  const {
-    availability,
-    key,
-    start,
+    isUserEnrolled,
     isEnrollable,
-    pacingType,
-    courseUuid,
-    seats,
-  } = activeCourseRun;
-
-  const sku = useMemo(
-    () => findHighestLevelSeatSku(seats),
-    [seats],
-  );
-  const isCourseStarted = useMemo(
-    () => hasCourseStarted(start),
-    [start],
-  );
-  const userEnrollment = useMemo(
-    () => findUserEnrollmentForCourse({ userEnrollments, key }),
-    [userEnrollments, key],
-  );
-  const enrollmentUrl = useCourseEnrollmentUrl({
-    catalogList,
-    enterpriseConfig,
-    key,
-    location,
-    offers,
-    sku,
-    subscriptionLicense,
-    userSubsidy,
+    isCourseStarted,
+    subsidyData,
   });
 
-  const EnrollLabel = props => (
+  const enrollLabel = props => (
```

Comment:
nit: `EnrollLabel` is capitalized because it's technically a functional component. It could abstracted out into its own file and/or function outside of `EnrollButton`.

Reply:
ah ok I will handle this i the next pr when I optimize enroll button label, thanks